lattice-controller/src/client_cache.rs

```diff
@@ -0,0 +1,224 @@
+    /// fallback credentials
+    pub(crate) async fn get_client(&self, lattice_id: &str) -> RpcResult<Client> {
+        let c = {
+            // Don't hold the read lock for the whole func
+            let lock = self.clients.read().await;
+            lock.get(lattice_id).cloned()
+        };
+        if let Some(c) = c {
+            self.record_access(lattice_id).await;
+            Ok(c)
+        } else {
+            let meta = {
+                // Dispose of lock as soon as we get what we need
+                let lock = self.meta.read().await;
+                lock.get(lattice_id).cloned()
+            };
+            if let Some(cfg) = meta {
+                let client = create_client(&cfg.config).await?;
+                self.store_client(lattice_id, client.clone()).await;
+                Ok(client)
+            } else {
+                Err(RpcError::NotInitialized(format!(
+                    "No client configuration for lattice {} stored",
+                    lattice_id
+                )))
+            }
+        }
+    }
+
+    async fn store_client(&self, lattice_id: &str, client: Client) {
+        let mut conns = self.clients.write().await;
+        conns.insert(lattice_id.to_string(), client);
+    }
+
+    async fn record_access(&self, lattice_id: &str) {
+        let mut meta = self.meta.write().await;
+        meta.entry(lattice_id.to_string()).and_modify(|e| e.touch());
+    }
+}
+
+async fn create_client(
+    config: &ConnectionConfig,
+) -> RpcResult<wasmcloud_control_interface::Client> {
+    let timeout = Duration::from_millis(config.timeout_ms);
+    let auction_timeout = Duration::from_millis(config.auction_timeout_ms);
+    let lattice_prefix = config.lattice_prefix.clone();
+    let conn = connect(config).await?;
+    let client = wasmcloud_control_interface::Client::new(
+        conn,
+        Some(lattice_prefix),
+        timeout,
+        auction_timeout,
+    );
+    Ok(client)
+}
+
+/// Create a new nats connection
+async fn connect(cfg: &ConnectionConfig) -> RpcResult<async_nats::Client> {
+    let cfg = cfg.clone();
+    let opts = match (cfg.auth_jwt, cfg.auth_seed) {
+        (Some(jwt), Some(seed)) => {
+            let key_pair = std::sync::Arc::new(
+                KeyPair::from_seed(&seed)
+                    .map_err(|e| RpcError::ProviderInit(format!("key init: {}", e)))?,
+            );
+            async_nats::ConnectOptions::with_jwt(jwt, move |nonce| {
+                let key_pair = key_pair.clone();
+                async move { key_pair.sign(&nonce).map_err(async_nats::AuthError::new) }
+            })
+        }
+        (None, None) => async_nats::ConnectOptions::default(),
+        _ => {
+            return Err(RpcError::InvalidParameter(
+                "must provide both jwt and seed for jwt authentication".into(),
+            ));
+        }
+    };
+    let url = cfg.cluster_uris.get(0).unwrap();
+    let conn = opts
+        .connect(url)
```

Comment:
Should we add an `event_callback` handler?

https://docs.rs/async-nats/0.18.0/async_nats/struct.ConnectOptions.html#method.event_callback

Reply:
you can use  `let conn = wasmbus_rpc::rpc_client::with_connection_event_logging(opts).connect(url)`
to log on callbacks